bundle/src/main/java/com/adobe/acs/commons/mcp/impl/processes/BrokenLinksReport.java

```diff
@@ -0,0 +1,211 @@
+     * @param htmlFields  lst of properties containing html
+     * @return stream containing extracted references
+     */
+    static Stream<String> collectPaths(Map.Entry<String, Object> property, Set<String> htmlFields) {
+        Object p = property.getValue();
+
+        Stream<String> stream;
+        if (p.getClass() == String[].class) {
+            stream = Arrays.stream((String[]) p);
+        } else if (p.getClass() == String.class){
+            stream = Stream.of((String) p);
+        } else {
+            stream = Stream.empty();
+        }
+        if (htmlFields.contains(property.getKey())) {
+            stream = stream.flatMap(val -> {
+                try {
+                    // parse html and extract links via underlying tagsoup library
+                    LinkContentHandler linkHandler = new LinkContentHandler();
+                    HtmlParser parser = new HtmlParser();
+                    parser.parse(new ByteArrayInputStream(val.getBytes("utf-8")), linkHandler, new Metadata(), new ParseContext());
+                    return linkHandler.getLinks().stream().map(Link::getUri);
+                } catch (Exception e) {
+                    return Stream.empty();
+                }
+            });
+        }
+        return stream;
+    }
+
+    /**
+     * Collect broken references from properties of the given resource
+     *
+     * @param resource      the resource to check
+     * @param regex         regex to to detect properties containing references. Set from @FormField
+     * @param skipList      properties to ignore. Set from @FormField
+     * @param htmlFields    field containing html .
+     * @return broken references keyed by property. The value is a List because a property can contain multiple links,
+     * e.g. if it is multivalued or it is html containing multiple links.
+     */
+    static Map<String, List<String>> collectBrokenReferences(Resource resource, Pattern regex, Set<String> skipList, Set<String> htmlFields) {
+
+        return resource.getValueMap().entrySet().stream()
+                .filter(entry -> !skipList.contains(entry.getKey()))
+                .collect(Collectors.toMap(
+                        entry -> resource.getPath() + "/" + entry.getKey(),
+                        entry -> {
+                            List<String> brokenPaths =  collectPaths(entry, htmlFields)
+                                    .filter(href -> regex.matcher(href).matches())
+                                    .filter(path -> ResourceUtil.isNonExistingResource(resource.getResourceResolver().resolve(path)))
+                                    .collect(Collectors.toList());
+                            return brokenPaths;
+                        })).entrySet().stream().filter(e -> !e.getValue().isEmpty())
+                .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
+     }
+        // access from unit tests
+    Map<String, EnumMap<REPORT, Object>> getReportData() {

```

Comment:
Needs to be a synchronized map otherwise race conditions will cause concurrent modification exceptions

Reply:
OK, I changed TreeMap to ConcurrentHashMap.  This way the pooled threads can safely put into it.